src/context/GlobalContext.tsx

```diff
@@ -0,0 +1,64 @@
+import React, {
+  useContext,
+  useReducer,
+  useEffect,
+  createContext,
+  useMemo,
+} from "react";
+import {
+  ContextActions,
+  GlobalState,
+  Types,
+} from "src/context/context.interface";
+import * as ActionCreators from "src/context/ActionCreators";
+import { rootReducer } from "src/context/reducer";
+import { AuthFormData } from "src/common/interface";
+import { onAuthStateChanged } from "src/modules/AuthService";
+import { User } from "src/modules/User";
+
+const initialState: GlobalState = {
+  loading: true,
+};
+const GlobalContext = createContext<{
+  state: GlobalState;
+  actions?: ContextActions;
+}>({
+  state: initialState,
+});
+
+export function useAuth() {
+  return useContext(GlobalContext);
+}
+
+export const GlobalProvider: React.FC = ({ children }) => {
+  const [state, dispatch] = useReducer(rootReducer, initialState);
+
+  const actions = useMemo<ContextActions>(
+    () => ({
+      signUp: (payload: AuthFormData) =>
+        ActionCreators.register(dispatch, payload),
+      signIn: (payload: AuthFormData) =>
+        ActionCreators.login(dispatch, payload),
+      signOut: () => ActionCreators.logout(dispatch),
+    }),
+    []
+  );
+
+  useEffect(() => {
+    const subscribeAuthChange = onAuthStateChanged((user) => {
+      // to avoid setting user if no name available
+      if (user && !user?.displayName) return;
```

Comment:
`user` should always have a `displayName`

We still need to subscribe even if `user` isn't set, to handle use-cases like `signIn` and `signUp`.

Reply:
this check is added because `user.updateProfile` triggers after `signUp` function so we don't need to set user until displayName is added.